### PR TITLE
Add cloud credentials management to the new profile

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- visit the new profile
- click on "Cloud credentials"
- you should see your cloud credentials

**Comment by huwshimi** [comment]
[Add credential management to profile](https://trello.com/c/I0EMclPt/139-add-credential-management-to-profile)

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8374/

**Comment by cassiocassio** [comment]
Profile appears and disappears in an instant, dropping me back to the canvas. Perhaps I'm building it wrong.

I see this in the console:

TypeError: Cannot read property 'displayName' of undefined
    at models.reduce (combo?app/init-pkg.js:23137)
    at Array.reduce (<anonymous>)
    at ProfileModelList._generateModels (combo?app/init-pkg.js:23113)
    at ProfileModelList.render (combo?app/init-pkg.js:23251)
    at finishClassComponent (combo?app/init-pkg.js:85692)
    at updateClassComponent (combo?app/init-pkg.js:85669)
    at beginWork (combo?app/init-pkg.js:86048)
    at performUnitOfWork (combo?app/init-pkg.js:88016)
    at workLoop (combo?app/init-pkg.js:88125)
    at HTMLUnknownElement.callCallback (combo?app/init-pkg.js:76742)
models.reduce @ combo?app/init-pkg.js:23137
_generateModels @ combo?app/init-pkg.js:23113
render @ combo?app/init-pkg.js:23251
finishClassComponent @ combo?app/init-pkg.js:85692
updateClassComponent @ combo?app/init-pkg.js:85669
beginWork @ combo?app/init-pkg.js:86048
performUnitOfWork @ combo?app/init-pkg.js:88016
workLoop @ combo?app/init-pkg.js:88125
callCallback @ combo?app/init-pkg.js:76742
invokeGuardedCallbackDev @ combo?app/init-pkg.js:76781
invokeGuardedCallback @ combo?app/init-pkg.js:76638
performWork @ combo?app/init-pkg.js:88243
scheduleUpdateImpl @ combo?app/init-pkg.js:88628
scheduleUpdate @ combo?app/init-pkg.js:88567
enqueueSetState @ combo?app/init-pkg.js:85089
ReactComponent.setState @ combo?app/init-pkg.js:93186
_fetchModelsCallback @ combo?app/init-pkg.js:23037
modelInfo @ combo?app/assets/javascripts/yui/arraylist/arraylist.js&app/assets/javascripts/yui/model-list/model-list.js&app/models/entity-extension.js&app/models/charm.js&app/models/bundle.js&app/assets/javascripts/yui/pluginhost-base/pluginhost-base.js&app/assets/javascripts/yui/pluginhost-config/pluginhost-config.js&app/assets/javascripts/yui/base-pluginhost/base-pluginhost.js&app/store/env/base.js&app/store/env/controller-api.js&app/store/env/api.js&app/assets/javascripts/yui/lazy-model-list/lazy-model-list.js&app/assets/javascripts/yui/datasource-local/datasource-local.js&app/assets/javascripts/yui/querystring-stringify-simple/querystring-stringify-simple.js&app/assets/javascripts/yui/io-base/io-base.js&app/assets/javascripts/yui/datasource-io/datasource-io.js&app/assets/javascripts/yui/plugin/plugin.js&app/assets/javascripts/yui/dataschema-base/dataschema-base.js&app/assets/javascripts/yui/json-stringify/json-stringify.js:4127
handler @ combo?app/assets/javascripts/yui/arraylist/arraylist.js&app/assets/javascripts/yui/model-list/model-list.js&app/models/entity-extension.js&app/models/charm.js&app/models/bundle.js&app/assets/javascripts/yui/pluginhost-base/pluginhost-base.js&app/assets/javascripts/yui/pluginhost-config/pluginhost-config.js&app/assets/javascripts/yui/base-pluginhost/base-pluginhost.js&app/store/env/base.js&app/store/env/controller-api.js&app/store/env/api.js&app/assets/javascripts/yui/lazy-model-list/lazy-model-list.js&app/assets/javascripts/yui/datasource-local/datasource-local.js&app/assets/javascripts/yui/querystring-stringify-simple/querystring-stringify-simple.js&app/assets/javascripts/yui/io-base/io-base.js&app/assets/javascripts/yui/datasource-io/datasource-io.js&app/assets/javascripts/yui/plugin/plugin.js&app/assets/javascripts/yui/dataschema-base/dataschema-base.js&app/assets/javascripts/yui/json-stringify/json-stringify.js:3999
dispatch_result @ combo?app/assets/javascripts/yui/arraylist/arraylist.js&app/assets/javascripts/yui/model-list/model-list.js&app/models/entity-extension.js&app/models/charm.js&app/models/bundle.js&app/assets/javascripts/yui/pluginhost-base/pluginhost-base.js&app/assets/javascripts/yui/pluginhost-config/pluginhost-config.js&app/assets/javascripts/yui/base-pluginhost/base-pluginhost.js&app/store/env/base.js&app/store/env/controller-api.js&app/store/env/api.js&app/assets/javascripts/yui/lazy-model-list/lazy-model-list.js&app/assets/javascripts/yui/datasource-local/datasource-local.js&app/assets/javascripts/yui/querystring-stringify-simple/querystring-stringify-simple.js&app/assets/javascripts/yui/io-base/io-base.js&app/assets/javascripts/yui/datasource-io/datasource-io.js&app/assets/javascripts/yui/plugin/plugin.js&app/assets/javascripts/yui/dataschema-base/dataschema-base.js&app/assets/javascripts/yui/json-stringify/json-stringify.js:3512
on_message @ combo?app/assets/javascripts/yui/arraylist/arraylist.js&app/assets/javascripts/yui/model-list/model-list.js&app/models/entity-extension.js&app/models/charm.js&app/models/bundle.js&app/assets/javascripts/yui/pluginhost-base/pluginhost-base.js&app/assets/javascripts/yui/pluginhost-config/pluginhost-config.js&app/assets/javascripts/yui/base-pluginhost/base-pluginhost.js&app/store/env/base.js&app/store/env/controller-api.js&app/store/env/api.js&app/assets/javascripts/yui/lazy-model-list/lazy-model-list.js&app/assets/javascripts/yui/datasource-local/datasource-local.js&app/assets/javascripts/yui/querystring-stringify-simple/querystring-stringify-simple.js&app/assets/javascripts/yui/io-base/io-base.js&app/assets/javascripts/yui/datasource-io/datasource-io.js&app/assets/javascripts/yui/plugin/plugin.js&app/assets/javascripts/yui/dataschema-base/dataschema-base.js&app/assets/javascripts/yui/json-stringify/json-stringify.js:3380
ws.onmessage @ combo?app/state/state.js&app/user/user.js&app/utils/github-ssh-keys.js&app/utils/statsd.js&app/jujulib/index.js&app/jujulib/charmstore.js&app/jujulib/bundleservice.js&app/jujulib/plans.js&app/jujulib/payment.js&app/jujulib/stripe.js&app/jujulib/terms.js&app/jujulib/reconnecting-websocket.js&app/jujulib/urls.js&app/jujulib/bakery.js:3802

**Comment by hatched** [comment]
@cassiocassio this is an existing issue in develop, I'll prepare a fix.

**Comment by hatched** [comment]
This branch will not be able to be properly QA'd until this fix has landed: https://github.com/juju/juju-gui/pull/3397

**Comment by huwshimi** [comment]
I've rebased this with the changes from #3397. Should be ready to try another QA.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8381/

File: jujugui/static/gui/src/app/components/profile/profile.js [hunk 1]
```diff
@@ -114,11 +134,27 @@ Profile.propTypes = {
...
+    revokeCloudCredential: PropTypes.func.isRequired,
+    updateCloudCredential: PropTypes.func.isRequired
+  }).isRequired,
+  controllerIsReady: PropTypes.func.isRequired,
+  controllerUser: PropTypes.string.isRequired,
   deployTarget: PropTypes.func.isRequired,
   destroyModels: PropTypes.func.isRequired,
   facadesExist: PropTypes.bool.isRequired,
   getModelName: PropTypes.func.isRequired,
+  initUtils: shapeup.shape({
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/profile/profile.js]: nice :)

**Comment by mitechie** [comment]
QA notes: (some of these are bugs and not necessarily with this branch. Making notes and adding @cassiocassio to start the conversation on what will end up as bugs/etc going forward)

    Credentials are stored securely on our servers and we will notify you by email whenever they are changed or deleted. You can see where they are used and manage or remove them via the account page.

This shows when clicking edit for each credential. It would appear this would be best handled once on the page and we don't need the sentence about how to manage/remove them as you're on that page currently. @cassiocassio 

During "Add" the google cloud image runs over the block on my 13" macbook screen. See attached image

credential-qa

The button experience during add seems off. There's two "Cancel" plus the "Change cloud" options here. Perhaps we just need the one cancel button?

cred-qa-2

When you click edit it feels like there should be a background color change on the row you're editing to indicate a bit of a selected state.

As a feature request it'd be great if we could tell on here how many models are using this credential and bonus points for which ones. 

Rather than "Provider" I think it'd be better to note it as a Cloud. In the CLI you specify adding a credential with a "Cloud Name" and a provider is an internal Juju concept vs a user one. 

    juju add-credential
    ERROR Usage: juju add-credential <cloud-name> [-f <credentials.yaml>]

**Comment by huwshimi** [comment]
Thanks heaps for the big QA @mitechie. I've added an issue with those comments over at #3398. I'm going to ship this branch to keep moving as it is all existing functionality.

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**